File: src/main/java/co/uk/isxander/lunarspoof/websocket/auth/packet/impl/CPacketEncryptionResponse.java
package co.uk.isxander.lunarspoof.websocket.auth.packet.impl;

import co.uk.isxander.lunarspoof.websocket.auth.LunarAuthWebSocket;
import co.uk.isxander.lunarspoof.websocket.auth.packet.LunarPacket;
import co.uk.isxander.xanderlib.utils.json.BetterJsonObject;
import net.minecraft.util.CryptManager;

import javax.crypto.SecretKey;
import java.security.PublicKey;
import java.util.Base64;

public class CPacketEncryptionResponse extends LunarPacket {

    private final byte[] secretHash;
    private final byte[] publicHash;

    public CPacketEncryptionResponse(SecretKey secretKey, PublicKey publicKey, byte[] keyHash) {
        this.secretHash = CryptManager.encryptData(publicKey, secretKey.getEncoded());
        this.publicHash = CryptManager.encryptData(publicKey, keyHash);
    }

    @Override
    public String getName() {
        return "CPacketEncryptionResponse";
    }

    @Override
    public void process(LunarAuthWebSocket ws) {
    }

    @Override
    public void processJson(BetterJsonObject json) {
        Base64.Encoder encoder = Base64.getUrlEncoder();
        json.addProperty("secretKey", new String(encoder.encode(secretHash)));
        json.addProperty("publicKey", new String(encoder.encode(publicHash)));
    }

}

File: src/main/kotlin/dev/isxander/lunarspoof/websocket/auth/packet/impl/CPacketEncryptionResponse.kt
package dev.isxander.lunarspoof.websocket.auth.packet.impl

import dev.isxander.lunarspoof.websocket.auth.LunarAuthWebSocket
import dev.isxander.lunarspoof.websocket.auth.packet.LunarPacket
import gg.essential.api.utils.JsonHolder
import net.minecraft.util.CryptManager
import java.security.PublicKey
import java.util.*
import javax.crypto.SecretKey

class CPacketEncryptionResponse(
    secretKey: SecretKey,
    publicKey: PublicKey,
    keyHash: ByteArray,
) : LunarPacket() {
    private val secretHash: ByteArray
    private val publicHash: ByteArray
    override val name: String = "CPacketEncryptionResponse"

    override fun process(ws: LunarAuthWebSocket) {}
    override fun processJson(json: JsonHolder) {
        val encoder = Base64.getUrlEncoder()
        json.put("secretKey", String(encoder.encode(secretHash)))
        json.put("publicKey", String(encoder.encode(publicHash)))
    }

    init {
        secretHash = CryptManager.encryptData(publicKey, secretKey.encoded)
        publicHash = CryptManager.encryptData(publicKey, keyHash)
    }
}